### ISSUE TITLE
[discuss] temp files and linters that use file name and file path info

### ISSUE THREAD

**Issue by braver** [SublimeLinter/SublimeLinter#1073]
#954

@sebastienbarre:

> > That said, especially in a unix-like context, programs that work on a file rarely care about that file's name. If a program receives a file to lint, it should usually just lint it unless the linter's author has been smoking something. 
> 
> @braver that's a bold assumption :) Case in point, SL4 broke a few of my linter configurations right away because it is passing a temporary filename now (@kaste was quick to fix one of them, thank you sir). I can think of many real scenarios. 
> 
> For example, you would want to ignore a whole sub-tree because it's legacy code that would require more work before getting validated as strictly as the rest of your source tree. Or you would want to ignore a specific file because it is boilerplate code that is created and maintained by the framework you are using (say, `create-react-app`, or some CakePHP console script).
>  
> Another example, in a classic Model-View-Controller architecture your controller classes often use snake case for methods (say, `api_set_price()`), whereas your model classes use camel case (say, `setPrice()`), and you rely on the filename to validate one way or the other (`FooController` for the controllers, vs. `Foo` or `FooModel` for the models).
> 
> Many popular frameworks nowadays are popular because they favor convention over configuration. Filenames are such a convention.
> 
> These exclusions or switches should live in (and be interpreted from) the version-controlled linter config file, not in every developer's favorite editor's linting framework config file :) And in that respect, SL3 was doing pretty great. It has been a bit rough with SL4, but I think we are getting somewhere :) With that being said, [specially crafted](https://github.com/SublimeLinter/SublimeLinter/issues/1066#issuecomment-370161075) temporary file names could have helped too...

@braver:

> > that's a bold assumption :)
> 
> Yeah I tend to do that. 😉 
> 
> Anyway, what you're saying basically means we need to reproduce the entire (relative) path to a file when we make a temporary file. Of course people using Windows will run into problems. And it generally feels inelegant to me.
> 
> In such cases, to me it sounds like, if where is file is and how it is named is apparently so important, temp files are not a good idea. Same with linters that take imports into account etc. That doesn't gel with temp files and the design of SublimeLinter. We could probably get a lot of it to work again, but the value-effort balance feels off to me. 
> 
> To me it makes more sense to bail out of this situation and only lint on-save using the real files.

**Comment by sebastienbarre** [comment]
> To me it makes more sense to bail out of this situation and only lint on-save using the real files.

That actually would be a good workaround for people currently experiencing these issues, until all Sublime linters are upgraded to support SL4. I would have no qualms changing from "background" to "load/save" linting as a compromise.

**Comment by braver** [comment]
Package Control doesn’t have version pinning. However, we introduced a version number that plugins can switch on to support both SL4 and SL3. SL3 will never be updated again, so you can install that manually.

**Comment by FichteFoll** [comment]
My first attempt would also be to make the linter only lint on save, once that is user-configurable (#244).

**Comment by braver** [comment]
tldr;

- We try our best to give linters the info they need, be it config, the original file name etc. This can usually be improved per plugin if needed.
- We use stdin (if available), or temp files, to allow on-the-fly linting. Some linters have provisions to allow for file-name or file-path related features in this situation, but if they don't there is nothing we can do. It's simply not possible to recreate a file's complete environment in a temp file.
- If a linter requires such file-name/path information to function completely, the plugin should not allow on-the-fly linting and declare itself real-file-only. Even if it's like an "enhanced mode", for consistency and predictability the plugin should in this case not run on-the-fly. A plugin could make it user-configurable if that's a good idea for that linter, but we're not seeing many cases where it is.
